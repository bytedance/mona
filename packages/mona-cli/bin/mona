#! /usr/bin/env node

const mona = require('../dist').default;
mona();
