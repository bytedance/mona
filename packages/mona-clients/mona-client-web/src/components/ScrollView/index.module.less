.container {
  position: relative;
  -webkit-overflow-scrolling: touch;
  height: 100%;
  width: 100%;
  max-height: inherit;

  &::-webkit-scrollbar {
    display: none;
  }
}

