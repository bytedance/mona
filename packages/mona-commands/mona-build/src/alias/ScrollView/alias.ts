import { bubbleEventsAlias, basePropAlias, baseComponentPropsMap as aliasMap } from '../prop';
import { ScrollViewAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  scrollX: aliasMap.scrollX,
  scrollY: aliasMap.scrollY,
  scrollIntoView: aliasMap.scrollIntoView,
  scrollWithAnimation: aliasMap.scrollWithAnimation,
  scrollLeft: aliasMap.scrollLeft,
  scrollTop: aliasMap.scrollTop,
  upperThreshold: aliasMap.upperThreshold,
  lowerThreshold: aliasMap.lowerThreshold,
  onScroll: aliasMap.onScroll,
  onScrollToLower: aliasMap.onScrollToLower,
  onScrollToUpper: aliasMap.onScrollToUpper,
};

export default alias;
