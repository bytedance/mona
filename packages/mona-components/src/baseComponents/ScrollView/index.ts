import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

export interface ScrollViewProps extends BaseProps {
  scrollX?: boolean;
  scrollY?: boolean;
  scrollWithAnimation?: boolean;
  upperThreshold?: number;
  lowerThreshold?: number;
  scrollTop?: number;
  scrollLeft?: number;
  scrollIntoView?: string;
  onScroll?: EventHandler;
  onScrollToUpper?: EventHandler;
  onScrollToLower?: EventHandler;
}

export const ScrollView: React.ComponentType<ScrollViewProps> = createComponent<ScrollViewProps>('scroll-view')
