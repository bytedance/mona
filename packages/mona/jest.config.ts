import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
