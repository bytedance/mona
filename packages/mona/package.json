{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
