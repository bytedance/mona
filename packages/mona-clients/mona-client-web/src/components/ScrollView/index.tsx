import React, { useLayoutEffect, useMemo, useRef } from 'react';
import { ScrollViewProps } from '@bytedance/mona';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { formatSyntheticEvent } from '../utils';

//t = current time
//b = start value
//c = change in value
//d = duration
function easeInOutQuad(t: number, b: number, c: number, d: number) {
  t /= d/2;
	if (t < 1) return c/2*t*t + b;
	t--;
	return -c/2 * (t*(t-2) - 1) + b;
};


function scrollToWithAnimation(element: HTMLDivElement, to: number, duration: number, vertical: boolean) {
    const start = vertical ? element.scrollTop : element.scrollLeft;
    const change = to - start;
    let currentTime = 0;
    const increment = 20;
        
    const animateScroll = function(){        
        currentTime += increment;
        const val = easeInOutQuad(currentTime, start, change, duration);
        if (vertical) {
          element.scrollTop = val;
        } else {
          element.scrollLeft = val;
        }
        if(currentTime < duration) {
            setTimeout(animateScroll, increment);
        }
    };
    animateScroll();
}

type onScrollCallback = (ele: Element) => void;

interface ScrollViewContextProps {
  registerScrollCallback: (callback: onScrollCallback) => void;
  unregisterScrollCallback: (callback: onScrollCallback) => void;
}

export const ScrollViewContext = React.createContext<ScrollViewContextProps | null>(null)

const ScrollView: React.FC<ScrollViewProps> = (props) => {
  const {
    children,
    scrollIntoView,
    upperThreshold,
    lowerThreshold,
    scrollWithAnimation,
    scrollTop,
    scrollLeft,
    scrollX,
    scrollY,
    onScroll,
    onScrollToUpper,
    onScrollToLower,
    ...restProps
  } = props;
  const { handleClassName, ...handleProps } = useHandlers(restProps)
  const scrollRef = useRef<HTMLDivElement | null>(null);
  const topPreRef = useRef(0);
  const leftPreRef = useRef(0);

  const cbRefs = useRef<onScrollCallback[]>([]);
  const context = useMemo(() => ({
    registerScrollCallback: (callback: onScrollCallback) => {
      cbRefs.current.push(callback)
    },
    unregisterScrollCallback: (callback: onScrollCallback) => {
      const index = cbRefs.current.indexOf(callback);
      cbRefs.current.splice(index, 1)
    }
  }), [])

  const overflow = `${scrollX ? 'auto' : 'hidden'} ${scrollY ? 'auto' : 'hidden'}`;

  const handleScroll = (e: React.SyntheticEvent) => {
    if (typeof onScroll === 'function') {
      onScroll(formatSyntheticEvent({ event: e }))
    }

    
    if (scrollRef.current) {
      cbRefs.current.forEach(cb => cb(scrollRef.current!));

      const top = scrollRef.current.scrollTop;
      const left = scrollRef.current.scrollLeft;
      const currentUpperThreshold = upperThreshold || 0;
      const currentLowerThreshold = lowerThreshold || 0;

      if (typeof onScrollToUpper === 'function') {
        if (scrollY && ((currentUpperThreshold > topPreRef.current && currentUpperThreshold <= top) || (currentUpperThreshold < topPreRef.current && currentUpperThreshold >= top))) {
          onScrollToUpper(formatSyntheticEvent({ event: e, type: 'scrolltoupper' }))
        }
        if (scrollX && ((currentUpperThreshold > leftPreRef.current && currentUpperThreshold <= left) || (currentUpperThreshold < leftPreRef.current && currentUpperThreshold >= left))) {
          onScrollToUpper(formatSyntheticEvent({ event: e, type: 'scrolltoupper' }))
        }  
      }

      if (typeof onScrollToLower === 'function') {
        if (scrollY) {
          const clientHeight = scrollRef.current.clientHeight;
          const scrollHeight = scrollRef.current.scrollHeight;
          const currentLower = scrollHeight - clientHeight - top;
          const prevLower = scrollHeight - clientHeight - topPreRef.current;
          if ((currentLowerThreshold > prevLower && currentLowerThreshold <= currentLower) || (currentLowerThreshold < prevLower && currentLowerThreshold >= currentLower)) {
            onScrollToLower(formatSyntheticEvent({ event: e, type: 'scrolltolower' }))
          }
        }
        if (scrollX) {
          const clientWidth = scrollRef.current.clientWidth;
          const scrollWidth = scrollRef.current.scrollWidth;
          const currentLower = scrollWidth - clientWidth - left;
          const prevLower = scrollWidth - clientWidth - leftPreRef.current;
          if ((currentLowerThreshold > prevLower && currentLowerThreshold <= currentLower) || (currentLowerThreshold < prevLower && currentLowerThreshold >= currentLower)) {
            onScrollToLower(formatSyntheticEvent({ event: e, type: 'scrolltolower' }))
          }
        }
      }

      topPreRef.current = top;
      leftPreRef.current = left;
    }
  }

  useLayoutEffect(() => {
    if (scrollY && scrollRef.current && scrollTop != null) {
      if (scrollWithAnimation) {
        scrollToWithAnimation(scrollRef.current, scrollTop, 300, true);
      } else {
        scrollRef.current.scrollTop = scrollTop;
      }
    }
  }, [scrollTop, scrollY, scrollWithAnimation])

  useLayoutEffect(() => {
    if (scrollX && scrollRef.current && scrollLeft != null) {
       if (scrollWithAnimation) {
        scrollToWithAnimation(scrollRef.current, scrollLeft, 300, false);
      } else {
        scrollRef.current.scrollLeft = scrollLeft;
      }
    }
  }, [scrollLeft, scrollX, scrollWithAnimation])

  useLayoutEffect(() => {
    if (scrollRef.current && scrollIntoView) {
      const $target = scrollRef.current.querySelector(`#${scrollIntoView}`)
      if ($target) {
        $target.scrollIntoView({ behavior: scrollWithAnimation ? 'smooth' : 'auto', block: 'start', inline: 'start' })
      }
    }
  }, [scrollIntoView, scrollWithAnimation])


  return (
    <div {...handleProps} className={handleClassName(styles.container)}>
      <div ref={scrollRef} className={styles.container} onScroll={handleScroll} style={{ overflow }}>
        <ScrollViewContext.Provider value={context}>
          {children}
        </ScrollViewContext.Provider>
      </div>
    </div>
  )
}

export default ScrollView;
