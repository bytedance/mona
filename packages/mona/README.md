# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
