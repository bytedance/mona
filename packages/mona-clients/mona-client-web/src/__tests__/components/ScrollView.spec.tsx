import React from 'react';
import { mount } from 'enzyme';
import ScrollView from '../../components/ScrollView';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { ScrollViewProps } from '@bytedance/mona';

describe('web component: ScrollView', () => {
  mountTest(ScrollView)
  handlerTest({ Component: ScrollView, firstTag: 'div' });

  it('should render correctly', () => {
    expect(mount(<ScrollView />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: ScrollViewProps = {
      scrollX: true,
      scrollY: true,
      scrollWithAnimation: true,
      upperThreshold: 10,
      lowerThreshold: 10,
      scrollTop: 10,
      scrollLeft: 10,
      scrollIntoView: 'hello',
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<ScrollView {...props}>hello</ScrollView>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
