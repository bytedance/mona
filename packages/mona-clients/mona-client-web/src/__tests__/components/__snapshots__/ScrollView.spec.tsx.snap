// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: ScrollView should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="container class-name  "
    style="color: red;"
  >
    <div
      class="container"
      style="overflow: auto auto;"
    >
      hello
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: ScrollView should render correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="container   "
  >
    <div
      class="container"
      style="overflow: hidden hidden;"
    />
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
